pkg/storagev2/testsuite/benchmark.go

```diff
@@ -0,0 +1,297 @@
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	return &pairedEntryGenerator{
+		keyGenerator:         newFullRandomKeyGenerator(start, n),
+		randomValueGenerator: makeRandomValueGenerator(r, *compressionRatio, *valueSize),
+	}
+}
+
+func newSequentialEntryGenerator(n int) entryGenerator {
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	return &pairedEntryGenerator{
+		keyGenerator:         newSequentialKeyGenerator(n),
+		randomValueGenerator: makeRandomValueGenerator(r, *compressionRatio, *valueSize),
+	}
+}
+
+type keyGenerator interface {
+	NKey() int
+	Key(i int) []byte
+}
+
+type reversedKeyGenerator struct {
+	keyGenerator
+}
+
+var _ keyGenerator = (*reversedKeyGenerator)(nil)
+
+func (g *reversedKeyGenerator) Key(i int) []byte {
+	return g.keyGenerator.Key(g.NKey() - i - 1)
+}
+
+func newReversedKeyGenerator(g keyGenerator) keyGenerator {
+	return &reversedKeyGenerator{keyGenerator: g}
+}
+
+type roundKeyGenerator struct {
+	keyGenerator
+}
+
+var _ keyGenerator = (*roundKeyGenerator)(nil)
+
+func (g *roundKeyGenerator) Key(i int) []byte {
+	return g.keyGenerator.Key(i % g.NKey())
+}
+
+func newRoundKeyGenerator(g keyGenerator) keyGenerator {
+	return &roundKeyGenerator{keyGenerator: g}
+}
+
+type predefinedKeyGenerator struct {
+	keys [][]byte
+}
+
+func (g *predefinedKeyGenerator) NKey() int {
+	return len(g.keys)
+}
+
+func (g *predefinedKeyGenerator) Key(i int) []byte {
+	if i >= len(g.keys) {
+		return g.keys[0]
+	}
+	return g.keys[i]
+}
+
+func newRandomKeyGenerator(n int) keyGenerator {
+	return &predefinedKeyGenerator{keys: newRandomKeys(n, hitKeyFormat)}
+}
+
+func newRandomMissingKeyGenerator(n int) keyGenerator {
+	return &predefinedKeyGenerator{keys: newRandomKeys(n, missingKeyFormat)}
+}
+
+func newFullRandomKeyGenerator(start, n int) keyGenerator {
+	return &predefinedKeyGenerator{keys: newFullRandomKeys(n, start, hitKeyFormat)}
+}
+
+func newSortedRandomKeyGenerator(n int) keyGenerator {
+	keys := newRandomKeys(n, hitKeyFormat)
+	sort.Slice(keys, func(i, j int) bool {
+		return bytes.Compare(keys[i], keys[j]) < 0
+	})
+	return &predefinedKeyGenerator{keys: keys}
+}
+
+func newSequentialKeyGenerator(n int) keyGenerator {
+	return &predefinedKeyGenerator{keys: newSequentialKeys(n, 0, hitKeyFormat)}
+}
+
+func maxInt(a int, b int) int {
+	if a >= b {
+		return a
+	}
+	return b
+}
+
+func doRead(b *testing.B, db storage.Store, g keyGenerator, allowNotFound bool) {
+	for i := 0; i < b.N; i++ {
+		key := g.Key(i)
+		item := &obj1{
+			Id: string(key),
+		}
+		err := db.Get(item)
+		switch {
+		case err == nil:
+		case allowNotFound && errors.Is(err, storage.ErrNotFound):
+		default:
+			b.Fatalf("%d: db get key[%s] error: %s\n", b.N, key, err)
+		}
+	}
+}
+
+type singularDBWriter struct {
+	db storage.Store
+}
+
+func (w *singularDBWriter) Put(key, value []byte) {
+	item := &obj1{
+		Id:  string(key),
+		Buf: value,
+	}
+	err := w.db.Put(item)
+	if err != nil {
+		panic(err)
```

Comment:
Instead of panic maybe we can use the t.Fatal by have a reference to the test in the writer. I think failing a test that way is better.

Reply:
fixed